From: Fabrice Olivero <f-olivero@ti.com>
Date: Mon, 11 Apr 2011 13:46:36 +0000 (+0530)
Subject: OMAP4 DSS HDMI: Support limited full range and aspect ratio in AVI InfoFrame
X-Git-Tag: android-2.6.35-2.3-omap4.13.1~26
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=b0adacf2efb812c9da10bc05ea0aac8439625e80

OMAP4 DSS HDMI: Support limited full range and aspect ratio in AVI InfoFrame

Change-Id: I0a300475905d1d6e39c5350bcb097986fdcd8f7b
Signed-off-by: Fabrice Olivero <f-olivero@ti.com>
---

diff --git a/arch/arm/plat-omap/hdmi_lib.c b/arch/arm/plat-omap/hdmi_lib.c
index 17d7745..f34f1b7 100644
--- a/arch/arm/plat-omap/hdmi_lib.c
+++ b/arch/arm/plat-omap/hdmi_lib.c
@@ -1645,11 +1645,36 @@ int hdmi_lib_enable(struct hdmi_config *cfg)
 	hdmi.avi_param.db1b_no_vert_hori_verthori = INFOFRAME_AVI_DB1B_NO;
 	hdmi.avi_param.db1s_0_1_2 = INFOFRAME_AVI_DB1S_0;
 	hdmi.avi_param.db2c_no_itu601_itu709_extented = INFOFRAME_AVI_DB2C_NO;
-	hdmi.avi_param.db2m_no_43_169 = INFOFRAME_AVI_DB2M_NO;
+
+	/* Support AR in AVI infoframe */
+	switch (cfg->video_format) {
+	/* 16:9 */
+	case 4:
+	case 5:
+	case 16:
+	case 19:
+	case 20:
+	case 31:
+	case 32:
+	case 39:
+		hdmi.avi_param.db2m_no_43_169 = INFOFRAME_AVI_DB2M_169;
+		break;
+	/* 4:3 */
+	case 1:
+	case 2:
+	case 6:
+	case 17:
+	case 21:
+	case 29:
+	case 35:
+	case 37:
+		hdmi.avi_param.db2m_no_43_169 = INFOFRAME_AVI_DB2M_43;
+		break;
+	}
+
 	hdmi.avi_param.db2r_same_43_169_149 = INFOFRAME_AVI_DB2R_SAME;
 	hdmi.avi_param.db3itc_no_yes = INFOFRAME_AVI_DB3ITC_NO;
 	hdmi.avi_param.db3ec_xvyuv601_xvyuv709 = INFOFRAME_AVI_DB3EC_XVYUV601;
-	hdmi.avi_param.db3q_default_lr_fr = INFOFRAME_AVI_DB3Q_DEFAULT;
 	hdmi.avi_param.db3sc_no_hori_vert_horivert = INFOFRAME_AVI_DB3SC_NO;
 	hdmi.avi_param.db4vic_videocode = cfg->video_format;
 	hdmi.avi_param.db5pr_no_2_3_4_5_6_7_8_9_10 = INFOFRAME_AVI_DB5PR_NO;
diff --git a/drivers/video/omap2/dss/hdmi.c b/drivers/video/omap2/dss/hdmi.c
index f45656b..b5a97dc 100644
--- a/drivers/video/omap2/dss/hdmi.c
+++ b/drivers/video/omap2/dss/hdmi.c
@@ -440,19 +440,15 @@ static void hdmi_configure_lr_fr(void)
 {
 	int ret = 0;
 	if (hdmi.mode == 0 || (hdmi.mode == 1 && hdmi.code == 1)) {
-		ret = hdmi_configure_lrfr(HDMI_FULL_RANGE, 0);
+		ret = hdmi_configure_lrfr(HDMI_FULL_RANGE, 1);
 		if (!ret)
 			dispc_setup_color_fr_lr(1);
 		return;
-	}
-	if (hdmi.lr_fr) {
-		ret = hdmi_configure_lrfr(HDMI_FULL_RANGE, hdmi.force_set);
-		if (!ret && !hdmi.force_set)
-			dispc_setup_color_fr_lr(1);
 	} else {
-		ret = hdmi_configure_lrfr(HDMI_LIMITED_RANGE, hdmi.force_set);
-		if (!ret && !hdmi.force_set)
+		ret = hdmi_configure_lrfr(HDMI_LIMITED_RANGE, 1);
+		if (!ret)
 			dispc_setup_color_fr_lr(0);
+		return;
 	}
 }
 
@@ -1186,6 +1182,8 @@ static int hdmi_power_on(struct omap_dss_device *dssdev)
 		}
 	}
 
+	hdmi_configure_lr_fr();
+
 	hdmi_lib_enable(&hdmi.cfg);
 
 	hdmi_configure_lr_fr();